fancy-calling-conventions.md

```diff
@@ -0,0 +1,175 @@
+
+### 10.  args and kwargs (needs strings, dicts, lists, iterators, destructors?)
+```
+def foo(* args, ** kwargs) -> None:
+	for arg in args:
+		print(arg)
+	for key, value in kwargs.items():
+		print(key)
+		print(value)
+foo(1, 2, 3, a = 4, b = 5, c = 6)
+
+“””
+1
+2
+3
+a
+4
+b
+5
+c
+6
+“””
+```
+Combining arbitrary arguments with keyword arguments. The convention follows that arbitrary arguments are defined before the keyword arguments. For this program to work, we treat arbitrary arguments as a list and keyword arguments as a dictionary. Finally, there would need to be support for iterators for the program to walk through these list and dictionary structures.
+
+## Tests:
+
+### Default arguments
+For default arguments, we will type check the literals defined at the function signature to make sure it matches the type of parameter. For function calls and method calls, we have to check the number of parameters passed in. The number of parameters passed in must be at least greater than the number of parameters without default values and less than the number of total parameters defined in the function signature. Then, those values in the function/method calls must be type checked.
+
+### Arbitrary arguments
+For arbitrary arguments, we need to make sure that all of the arguments passed in matches with the type defined at the parameter.
+
+### Keyword arguments
+For testing keyword arguments, we will switch the argument positions but ensure that the output of the function stays the same.
+
+### Regular arguments
+To ensure the basic functionality of method calls and function calls are still intact, we will write the exact program without the new calling conventions and make sure the outputs are correct.
+
+## New AST forms:
+
+To account for parameters with default arguments, we would need to include a new pattern for the `Parameter` type in `ast.ts`. The new pattern should include a field for the default value.
+```
+export type Parameter<A>  =
+| { a?: A, tag: “reg”, name: string, type: Type }
+| { a?: A, tag: “default”, name: string, type: Type, value: Value }
+```
+
+Additionally, we would need to include a field in both call and method call expressions to account for the inclusion of keyword arguments. The `kwargs` field should correspond to an array of arrays of argument names followed by their values.
+```
+export type Expr<A> =
+| { a?: A, tag: "call", name: string, args: Array<Expr<A>>, kwargs: Array<[string, Expr<A>]> }
+| { a?: A, tag: "method-call", obj: Expr<A>, method: string, args: Array<Expr<A>>, kwargs: Array<[string, Expr<A>]> }
+```
+
+## New functions, datatypes, and/or files:
+
+In our global environment, we would need to add another map for functions which would keep track of the parameters names and its corresponding default values. We need to keep track of these fields in order to properly populate the arguments of function calls during compile time.
```

Comment:
As a member of the closures (as well as first-class functions and lambda expression) group, I would say that it seems very difficult to know the number of arguments of a callable at compile time when involving default arguments and first-class functions. Consider the following example:

``` python
f0: Callable[[], int]

# assuming there are 3 callables of type below:
funcbody: Callable[[int], int] = None
expr1: Callable[[], int] = None
cond1: Callable[[], bool] = None

# Above 3 callables are somehow assigned to non-trivial functions/lambda expressions

def f(x:int = expr1())-> int:
    return funcbody(x)

def g()-> int:
    return 1

# maybe some code here that changes the output of expr1() before and after it.  

if cond1():
    # Although `f0` is a `Callable[[], int]` without optional args which is not of  the identical type of `f`,
    # It is very boring if we decide to not support this kind of conformance of types
    f0 = f  
else:
    f0 = g

f0()
```

Reply:
Thinking about ChocoPy, I think it is probably in our best interests to restrict what gets put inside the default arguments.
I'm thinking of default arguments similar to the init statements in the previous PAs - that is, we should only be able to initialize a variable with a literal value.

@danielytan

Reply:
Thanks for your reply. The restriction will probably make life easier. We can come back to the discussion of storing parameters' default values after our group settling down the representation of callables. You are no more welcome to keep an eye and give suggestions on our PR thread if you wish. https://github.com/ucsd-cse231-w21/chocopy-wasm-compiler/pull/23#discussion_r584003981

Reply:
@DDEle 
Sorry for the late reply, but we are trying to do without the restriction.
We have a discussion going on in this thread [here](https://github.com/ucsd-cse231-w21/chocopy-wasm-compiler/pull/51#discussion_r587713692). Feel free to take a look.